File: keys-and-rooms/keys-and-rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        Deque<Integer> q = new ArrayDeque<>();
        q.offer(0);
        
        while (!q.isEmpty()) {
            int visit = q.poll();
            visited[visit] = true;
            for(int i = 0; i < rooms.get(visit).size(); i++) {
                if (!visited[rooms.get(visit).get(i)]) {
                    q.offer(rooms.get(visit).get(i));
                }
            }
        }

        for (int i = 0; i < visited.length; i++) {
            if (!visited[i]) return false;
        }
        return true;
    }
}

File: 871-keys-and-rooms/keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        visited = [0] * n
        def dfs(room):
            visited[room] = 1
            keys = rooms[room]
            for key in keys:
                if not visited[key]:
                    dfs(key)
        dfs(0)
        return sum(visited) == n